In [2]:
#import libraries
import pandas as pd
import numpy as np

In [3]:
#read users table
users = pd.read_json("json_data/lucid_table_users.json")
users.set_index('id',inplace=True)

In [4]:
#drop irrelevant columns
users.drop(['image','password','remember_token'],axis = 1, inplace=True)

In [5]:
#views users head
users.head()

created_at                         email                   name  \
id                                                                            
1  2019-08-01 17:57:22   oluwaseyieniayomi@gmail.com     Eniayomi Oluwaseyi   
2  2019-08-01 18:05:58           okoelijah@gmail.com          Elijah Okokon   
3  2019-08-01 20:55:39           jeff.ogah@gmail.com           Jeffrey Ogah   
4  2019-08-02 16:52:47     nathanoluwaseyi@gmail.com  Oluwaseyi Oluwapelumi   
5  2019-08-03 11:33:51  paulchibiukeigweze@gmail.com                   PoRH   

   provider   provider_id                                          short_bio  \
id                                                                             
1    google  1.008922e+20               Software Developer | DevOPs Engineer   
2    google  1.055840e+20                                      Web Developer   
3    google  1.012351e+20  Front End Developer | React Developer | Mentor...   
4    google  1.072335e+20  | Software Developer | DevOps Engineer | @linu...   
5    google  1.034316e+20            I Am lamar and you don't think am real?   

            updated_at   username  
id                                 
1  2019-08-01 17:57:22   eniayomi  
2  2019-08-01 18:05:58    DMatrix  
3  2019-08-01 20:55:39  jeff.ogah  
4  2019-08-02 16:52:47     nathan  
5  2019-08-03 11:33:51      lamar

In [7]:
#read users post
posts = pd.read_json("json_data/lucid_table_posts.json")
posts.set_index('id',inplace=True)

In [8]:
#drop irrelevant columns
posts.drop(['status_id','action','post_id'], axis = 1, inplace=True)

In [9]:
#view posts head
posts.head()

content          created_at  \
id                                                                          
1   I learnt how to use the table tag as i have us... 2019-08-25 19:39:01   
2    I am on this journey with start.ng, and here... 2019-08-25 19:44:36   
4   I have not been attending classes on the HNG c... 2019-08-25 19:50:33   
6   My journey on **StartNG** pre-internship progr... 2019-08-25 19:51:36   
7    A Summary on The “idongesit.html� CV, It... 2019-08-25 19:51:39   

   image                                   slug        tags  \
id                                                            
1      1  what-i-have-learnt-so-far-on-html-985        None   
2      1                   html-begins-here-4d6  Technology   
4      1            my-laziness-in-the-open-029        None   
6      1                          my-task-2-649        None   
7      1                             task-2-7e5        None   

                                title          updated_at  user_id  
id                                                                  
1   What i have learnt so far on HTML 2019-08-28 11:30:00     2077  
2                    HTML BEGINS HERE 2019-08-28 11:30:00     1719  
4             My Laziness In The Open 2019-08-28 11:30:00     1310  
6                           MY TASK 2 2019-08-28 11:30:00     1787  
7                              Task 2 2019-08-28 11:30:00      167

In [10]:
posts.user_id.nunique()

308

In [12]:
#read contact table
contact= pd.read_json("json_data/lucid_table_contact_settings.json")
contact.set_index('id',inplace=True)

In [13]:
#drop NaN
contact.dropna(axis = 1,inplace=True)

In [14]:
#view head
contact.head()

email  user_id
id                                      
1   emmanuelezenwigbo@gmail.com       78
2            dward009@gmail.com       90
3             apeji93@gmail.com       53
4           ireabueke@gmail.com      133
5           bindas.fs@gmail.com       92

In [17]:
#read following table
following = pd.read_json("json_data/lucid_table_following.json")

In [18]:
following.head()

follower_id  my_id  status
0            6      3       1
1            3      6       1
2            2      3       1
3            7      3       1
4            2      7       1

In [19]:
#read notifications table
notifications = pd.read_json("json_data/lucid_table_notifications.json")

In [20]:
#drop NaN
notifications.dropna(axis = 1,inplace=True)
notifications.drop(['id'], axis = 1, inplace=True)

In [21]:
notifications.head()

action          created_at  sender_id  status       type  \
0   Followed 2019-08-26 09:43:56          4       1  Following   
1   Followed 2019-08-26 09:48:21          1       1  Following   
2   Followed 2019-08-26 09:51:39          3       0  Following   
3   Followed 2019-08-26 11:32:46          1       1  Following   
4  Commented 2019-08-26 15:13:17          1       1       Post   

           updated_at  user_id  
0 2019-08-26 18:20:28        7  
1 2019-08-26 13:05:42        4  
2 2019-08-26 09:51:39       45  
3 2019-08-26 13:05:42        4  
4 2019-08-26 15:13:25        4

In [22]:
notifications.user_id.nunique()

25

In [23]:
#read thoughts table
thoughts = pd.read_json("json_data/lucid_table_thoughts.json")
thoughts.set_index('id',inplace=True)

In [24]:
thoughts.head()

content          created_at  \
id                                                                          
1                                            hhhhhhhh 2019-08-26 13:30:26   
2                       Can't believe I'm learning ML 2019-08-26 13:57:35   
3   - what you have learned so far\r\nI've learned... 2019-08-26 14:35:41   
4                                          Hey people 2019-08-26 15:12:07   
5                                                nice 2019-08-26 15:24:55   

            updated_at  user_id  
id                               
1  2019-08-26 13:30:26     2195  
2  2019-08-26 13:57:35      148  
3  2019-08-26 14:35:41     2212  
4  2019-08-26 15:12:07        1  
5  2019-08-26 15:24:55     2223

In [25]:
#merging posts and contact table
df1 = pd.merge(posts,contact,on='user_id')
df1.head()

content          created_at  \
0  Using the Hyper Text Markup Language (HTML) ha... 2019-08-25 19:54:07   
1  The screenshot of my HTML CV attached to my la... 2019-08-25 20:06:38   
2  <p>Using the Hyper Text Markup Language (HTML)... 2019-08-27 11:14:33   
3  <p>So far, have learnt how to optimally use Sl... 2019-08-27 11:14:33   
4                                       lucid why na 2019-08-25 20:04:09   

                                 image                    slug  tags  \
0                                    1  my-journey-on-html-ce5  None   
1                                 None         please-note-4f4  None   
2  topejoan2/images/img-q8lju377v8.png  my-journey-on-html-ce5         
3                                       my-learning-so-far-56f         
4                                 None                 why-8be  None   

                title          updated_at  user_id                     email  
0  My Journey on HTML 2019-08-28 11:30:34      329       topejoan2@gmail.com  
1         Please note 2019-08-28 11:30:01      329       topejoan2@gmail.com  
2  My Journey on HTML 2019-08-28 11:30:34      329       topejoan2@gmail.com  
3  My learning so far 2019-08-28 11:30:34      329       topejoan2@gmail.com  
4                 why 2019-08-28 11:30:01      942  enoekanem.vibe@gmail.com

In [26]:
#merging posts and notifications
df2 = pd.merge(posts,notifications,on='user_id')
df2.drop(['tags','created_at_x','updated_at_x','created_at_y','updated_at_y'], axis = 1, inplace = True)

In [27]:
#merging posts and thoughts
df3 = pd.merge(posts,thoughts,on='user_id')
df3.head()

content_x        created_at_x  \
0  I was scrolling through my tweets the other da... 2019-08-26 09:06:06   
1  **WHAT HAVE YOU LEARNED SO FAR?**\n\n  \n\nI'v... 2019-08-26 14:41:21   
2  I joined a newbie and I'm really ecstatic to b... 2019-08-26 19:34:31   
3                                                Hey 2019-08-26 15:10:44   
4                                                Hey 2019-08-26 15:10:44   

  image                                         slug  tags  \
0  None  coding101-through-the-eye-of-a-beginner-6e6  None   
1  None                  what-ive-learned-so-far-6d6  None   
2     1                               using-html-dd2  None   
3  None                    good-afternoon-people-288  None   
4  None                    good-afternoon-people-288  None   

                                      title        updated_at_x  user_id  \
0  CODING101: THROUGH THE EYE OF A BEGINNER 2019-08-28 11:30:07     2195   
1                  WHAT I'VE LEARNED SO FAR 2019-08-28 11:30:12     2212   
2                                USING HTML 2019-08-28 11:30:17     2212   
3                     Good Afternoon People 2019-08-28 11:30:12        4   
4                     Good Afternoon People 2019-08-28 11:30:12        4   

                                           content_y        created_at_y  \
0                                           hhhhhhhh 2019-08-26 13:30:26   
1  - what you have learned so far\r\nI've learned... 2019-08-26 14:35:41   
2  - what you have learned so far\r\nI've learned... 2019-08-26 14:35:41   
3                                         Hey people 2019-08-26 22:29:07   
4                                                Hey 2019-09-05 17:19:25   

         updated_at_y  
0 2019-08-26 13:30:26  
1 2019-08-26 14:35:41  
2 2019-08-26 14:35:41  
3 2019-08-26 22:29:07  
4 2019-09-05 17:19:25

In [28]:
#merge users and contact
df4 = pd.merge(users,contact, on='email')
df4.head()

created_at                   email            name provider  \
0 2019-08-11 11:36:15   idriseun222@gmail.com     Lawal Idris   google   
1 2019-08-18 21:49:58       apeji93@gmail.com     Yossi Apeji   google   
2 2019-08-18 21:50:35      akpos008@gmail.com  wanapere akpos   google   
3 2019-08-18 21:50:53   giftomodu20@gmail.com      Gift Omodu   google   
4 2019-08-18 21:51:31  josephachonu@gmail.com   Joseph Achonu   google   

    provider_id                                          short_bio  \
0  1.149969e+20                                  Software Engineer   
1  1.119468e+20  I am a solution's architect and Economist look...   
2  1.065443e+20                                Cool and easy going   
3  1.116077e+20  Back-end Developer...\r\n\r\nError\r\n#404\r\n...   
4  1.046430e+20  A Tech ethusiast and a strong believer in educ...   

           updated_at      username  user_id  
0 2019-08-11 11:36:15   idriseun222       19  
1 2019-08-18 21:49:58       apeji93       53  
2 2019-08-18 21:50:35      akpos008       61  
3 2019-08-18 21:50:53   giftomodu20       62  
4 2019-08-18 21:51:31  josephachonu       70

In [29]:
df2.groupby('title')['action'].count().sort_values(ascending = False).head()

title
CEO -- Story so Far                                             105
BUILD AROUND YOUR DISCOVERY MECHANISM                            42
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO     35
DON'T ASK WHY YOU WERE REJECTED                                  35
THE SUNFLOWER                                                    35
Name: action, dtype: int64

From above we can see that CEO -- Story so Far has the most actions i.e commented or followed 

In [30]:
users.head()

created_at                         email                   name  \
id                                                                            
1  2019-08-01 17:57:22   oluwaseyieniayomi@gmail.com     Eniayomi Oluwaseyi   
2  2019-08-01 18:05:58           okoelijah@gmail.com          Elijah Okokon   
3  2019-08-01 20:55:39           jeff.ogah@gmail.com           Jeffrey Ogah   
4  2019-08-02 16:52:47     nathanoluwaseyi@gmail.com  Oluwaseyi Oluwapelumi   
5  2019-08-03 11:33:51  paulchibiukeigweze@gmail.com                   PoRH   

   provider   provider_id                                          short_bio  \
id                                                                             
1    google  1.008922e+20               Software Developer | DevOPs Engineer   
2    google  1.055840e+20                                      Web Developer   
3    google  1.012351e+20  Front End Developer | React Developer | Mentor...   
4    google  1.072335e+20  | Software Developer | DevOps Engineer | @linu...   
5    google  1.034316e+20            I Am lamar and you don't think am real?   

            updated_at   username  
id                                 
1  2019-08-01 17:57:22   eniayomi  
2  2019-08-01 18:05:58    DMatrix  
3  2019-08-01 20:55:39  jeff.ogah  
4  2019-08-02 16:52:47     nathan  
5  2019-08-03 11:33:51      lamar

In [31]:
posts.head()

content          created_at  \
id                                                                          
1   I learnt how to use the table tag as i have us... 2019-08-25 19:39:01   
2    I am on this journey with start.ng, and here... 2019-08-25 19:44:36   
4   I have not been attending classes on the HNG c... 2019-08-25 19:50:33   
6   My journey on **StartNG** pre-internship progr... 2019-08-25 19:51:36   
7    A Summary on The “idongesit.html� CV, It... 2019-08-25 19:51:39   

   image                                   slug        tags  \
id                                                            
1      1  what-i-have-learnt-so-far-on-html-985        None   
2      1                   html-begins-here-4d6  Technology   
4      1            my-laziness-in-the-open-029        None   
6      1                          my-task-2-649        None   
7      1                             task-2-7e5        None   

                                title          updated_at  user_id  
id                                                                  
1   What i have learnt so far on HTML 2019-08-28 11:30:00     2077  
2                    HTML BEGINS HERE 2019-08-28 11:30:00     1719  
4             My Laziness In The Open 2019-08-28 11:30:00     1310  
6                           MY TASK 2 2019-08-28 11:30:00     1787  
7                              Task 2 2019-08-28 11:30:00      167

In [32]:
df2.drop(['content','slug','image'], axis = 1, inplace=True)

In [33]:
df2.head()

title  user_id     action  sender_id  status  \
0      What you have learned so far     2191  Commented       2191       0   
1      What you wish to still learn     2191  Commented       2191       0   
2  What you wish should be improved     2191  Commented       2191       0   
3             Good Afternoon People        4   Followed          1       1   
4             Good Afternoon People        4   Followed          1       1   

        type  
0       Post  
1       Post  
2       Post  
3  Following  
4  Following

In [34]:
df2.action.unique()

array(['Commented', 'Followed', 'Like', 'Replied', 'Love'], dtype=object)

In [35]:
def status(text):
    if text == 'Commented':
        return 1.0
    elif text == 'Followed':
        return 2.0
    elif text == 'Like':
        return 3.0
    elif text == 'Replied':
        return 4.0
    elif text == 'Love':
        return 5.0

In [36]:
df2['rating'] = df2['action'].apply(status)

In [37]:
df2.head()

title  user_id     action  sender_id  status  \
0      What you have learned so far     2191  Commented       2191       0   
1      What you wish to still learn     2191  Commented       2191       0   
2  What you wish should be improved     2191  Commented       2191       0   
3             Good Afternoon People        4   Followed          1       1   
4             Good Afternoon People        4   Followed          1       1   

        type  rating  
0       Post     1.0  
1       Post     1.0  
2       Post     1.0  
3  Following     2.0  
4  Following     2.0

In [38]:
#create a pivot table
df = df2.pivot_table(index = 'sender_id', columns = 'title', values = 'status', aggfunc='mean')

In [39]:
df.head()

title      A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO  \
sender_id                                                                 
1                                                   0.333333              
2                                                   0.888889              
3                                                   1.000000              
4                                                   0.888889              
6                                                        NaN              

title      ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?  Am ok  \
sender_id                                                           
1                                                      0.0    NaN   
2                                                      NaN    NaN   
3                                                      1.0    NaN   
4                                                      1.0    NaN   
6                                                      NaN    NaN   

title      BUILD AROUND YOUR DISCOVERY MECHANISM  CEO -- Story so Far  \
sender_id                                                               
1                                       0.200000                  NaN   
2                                       0.888889                  NaN   
3                                       1.000000                  0.0   
4                                       0.909091                  0.0   
6                                            NaN                  NaN   

title      Class Day 1  DO YOU NEED A CO-FOUNDER?  \
sender_id                                           
1                  NaN                        0.0   
2                  NaN                        NaN   
3                  0.0                        1.0   
4                  0.5                        1.0   
6                  1.0                        NaN   

title      DON'T ASK WHY YOU WERE REJECTED  EVALUATING STARTUP IDEAS  First  \
sender_id                                                                     
1                                 0.333333                       0.0    1.0   
2                                 0.888889                       NaN    NaN   
3                                 1.000000                       1.0    1.0   
4                                 0.888889                       1.0    NaN   
6                                      NaN                       NaN    NaN   

title      ...  asda  html test  jbhjkhkbb  ochem  second  third  \
sender_id  ...                                                     
1          ...   NaN   1.000000        NaN    NaN     1.0    1.0   
2          ...   NaN   0.888889        NaN    NaN     NaN    NaN   
3          ...   NaN   1.000000        0.0    NaN     1.0    1.0   
4          ...   NaN   0.857143        NaN    NaN     NaN    NaN   
6          ...   NaN        NaN        NaN    NaN     NaN    NaN   

title      this is a test post  this is mine  what have done so far on HTML  \
sender_id                                                                     
1                     1.000000      1.000000                            NaN   
2                     0.888889      0.888889                            NaN   
3                     1.000000      1.000000                            NaN   
4                     0.857143      0.857143                            NaN   
6                          NaN           NaN                            NaN   

title      with picture  
sender_id                
1              1.000000  
2              0.888889  
3              1.000000  
4              0.857143  
6                   NaN  

[5 rows x 55 columns]

In [40]:

x = df['CEO -- Story so Far']

In [41]:
#finding the correlation of other articles with CEO--story so far article
corr = df.corrwith(x)

C:\Users\BUCHI\Anaconda3\lib\site-packages\numpy\lib\function_base.py:2392: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar)
C:\Users\BUCHI\Anaconda3\lib\site-packages\numpy\lib\function_base.py:2326: RuntimeWarning: divide by zero encountered in true_divide
  c *= np.true_divide(1, fact)


In [42]:
corr

title
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO         NaN
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                      NaN
Am ok                                                                NaN
BUILD AROUND YOUR DISCOVERY MECHANISM                                NaN
CEO -- Story so Far                                             1.000000
Class Day 1                                                     0.522233
DO YOU NEED A CO-FOUNDER?                                            NaN
DON'T ASK WHY YOU WERE REJECTED                                      NaN
EVALUATING STARTUP IDEAS                                             NaN
First                                                                NaN
Good Afternoon People                                                NaN
HNG Pre-internship                                                   NaN
HOW TO AVOID BEING A ONE-HIT WONDER                                  NaN
Hello New Followers                          

Intuitively, we can see that CEO-- story so far is correlated with the titles with 0.522233 correlation values

In [43]:
#correlating the columns using pearson formula
matrix_corr = df.corr(method='pearson')

In [44]:
matrix_corr.head()

title                                               A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO  \
title                                                                                                              
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                                           1.000000              
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                              0.955072              
Am ok                                                                                             NaN              
BUILD AROUND YOUR DISCOVERY MECHANISM                                                        0.995061              
CEO -- Story so Far                                                                               NaN              

title                                               ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?  \
title                                                                                                 
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                                         0.955072   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                            1.000000   
Am ok                                                                                           NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                                                      0.984108   
CEO -- Story so Far                                                                             NaN   

title                                               Am ok  \
title                                                       
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...    NaN   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?       NaN   
Am ok                                                 NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                 NaN   
CEO -- Story so Far                                   NaN   

title                                               BUILD AROUND YOUR DISCOVERY MECHANISM  \
title                                                                                       
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                               0.995061   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                  0.984108   
Am ok                                                                                 NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                                            1.000000   
CEO -- Story so Far                                                                   NaN   

title                                               CEO -- Story so Far  \
title                                                                     
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                  NaN   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                     NaN   
Am ok                                                               NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                               NaN   
CEO -- Story so Far                                                 1.0   

title                                               Class Day 1  \
title                                                             
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...    -0.519170   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?       -0.866025   
Am ok                                                       NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                 -0.520205   
CEO -- Story so Far                                    0.522233   

title                                               DO YOU NEED A CO-FOUNDER?  \
title                                                                           
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                   0.955072   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                      1.000000   
Am ok                                                            

In [199]:
#posts in which the sender 1 acted on by either liking, loving, commenting and so on
df.iloc[0].dropna()

title
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO    0.333333
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                 0.000000
BUILD AROUND YOUR DISCOVERY MECHANISM                           0.200000
DO YOU NEED A CO-FOUNDER?                                       0.000000
DON'T ASK WHY YOU WERE REJECTED                                 0.333333
EVALUATING STARTUP IDEAS                                        0.000000
First                                                           1.000000
Good Afternoon People                                           1.000000
HOW TO AVOID BEING A ONE-HIT WONDER                             0.000000
SHALLOW THINKING                                                0.000000
TEst                                                            1.000000
THE SUNFLOWER                                                   0.333333
TWO YEARS IN THE TRENCHES BUILDING HOTELS.NG                    0.000000
This is lucid                                

In [45]:
#creating a series for sender corr
sender_corr = pd.Series()

sender_id = 1

for post in df.iloc[sender_id-1].dropna().index:
    corr_list = matrix_corr[post].dropna()*df.iloc[sender_id-1][post]
    sender_corr = sender_corr.append(corr_list)

In [46]:
#grouping user correlations by the index and summing up
sender_corr = sender_corr.groupby(sender_corr.index).sum()

In [47]:
#create an empty list for posts
post_list = []

for i in range(len(df.iloc[sender_id-1].dropna().index)):
    if df.iloc[sender_id-1].dropna().index[i] in sender_corr:
        post_list.append(df.iloc[sender_id-1].dropna().index[i])
    else:
        pass
sender_corr = sender_corr.drop(post_list)

In [48]:
print('Hi, based on the post that you reacted to: \n')
for i in df.iloc[sender_id-1].dropna().index:
    print(i)
print('\n I would suggest you read these posts \n')
for i in sender_corr.sort_values(ascending= False).index[:5]:
    print(i)

Hi, based on the post that you reacted to: 

A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?
BUILD AROUND YOUR DISCOVERY MECHANISM
DO YOU NEED A CO-FOUNDER?
DON'T ASK WHY YOU WERE REJECTED
EVALUATING STARTUP IDEAS
First
Good Afternoon People
HOW TO AVOID BEING A ONE-HIT WONDER
SHALLOW THINKING
TEst
THE SUNFLOWER
TWO YEARS IN THE TRENCHES BUILDING HOTELS.NG
This is lucid
Took a long while
Welcome to Lucid
YOUR HIGH IQ WILL KILL YOUR STARTUP
Your face PHP script
html test
second
third
this is a test post
this is mine
with picture

 I would suggest you read these posts 

Working round the clock on Lucid
Testing tags and really long posts
Testing image
Test
Stressful Day
